pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTModifierNode.java

```diff
@@ -4,14 +4,14 @@
 
 package net.sourceforge.pmd.lang.apex.ast;
 
-import com.google.summit.ast.Node;
+import com.google.summit.ast.modifier.Modifier;
 import net.sourceforge.pmd.annotation.InternalApi;
 
-public class ASTModifierNode extends AbstractApexNode<Node> implements AccessNode {
+public class ASTModifierNode extends AbstractApexNode<ListAdapter<Modifier>> implements AccessNode {
```

Comment:
Like `AbstractApexNode`, [`ApexNode`](https://github.com/eklimo/pmd/blob/summit/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexNode.java#L41) is designed to wrap only one node. Could you clarify how using `ApexNode` directly would help here?

Though you're right, this solution is a bit of a hack, I think it has some validity. One of my first proposed solutions for the many-to-one modifier node problem involved creating a new Summit node within PMD, e.g. `ModifierGroup`, that would have a `List<Modifier>` as a property and would be wrapped by `ASTModifierNode`. This way, any `List<Modifier>` that comes up can appropriately be converted into a `ASTModifierNode` that wraps a `ModifierGroup` that holds the `List<Modifier>`. So far, it doesn't seem like a hack to me - more a way to adapt the Summit AST as needed to align more with PMD's.

However, I saw that this structure could be generalized beyond just modifiers, so it naturally turned into `ListAdapter`, which is the fake node that works for any many-to-one mapping.

(Of course, if you have an issue with the fake node even in the `ModifierGroup` example, this isn't relevant.)

Reply:
The only part of the `ApexNode<T>` that involves the generic parameter is `@Deprecated T getNode`. If there are any external callers, they're already broken by replacing Jorje and changing T, right? I guess I was imagining that we need not try to fake a 1:1 Summit mapping for the benefit of this (best case: unused, worst case: already broken) method. For example, throwing an exception could be a fine placeholder for external callers. With regard to the generic parameter `T`, it could be instantiated with some dummy type -- until some a future refactoring eliminates it.

What we do internally is I guess where our thinking diverged. If we assume that the mapping between Summit and PMD is entirely internal and under our control, is there any value in faking 1:1 Summit nodes? This seems a bit convoluted to me. Skipping or splitting `AbstractApexNode` would eliminate the internal assumption that there is a `this.node`, and you could instead keep an internal `List<Node>` or whatever is appropriate.

Reply:
Thank you for elaborating. You're right, since `T` has already changed in `ApexNode`, callers of `getNode` are already broken. For example - and I never thought to compile/test other PMD modules, since I assumed they were all independent - it looks like something in [`pmd-visualforce`](https://github.com/pmd/pmd/blob/master/pmd-visualforce/src/main/java/net/sourceforge/pmd/lang/vf/ApexClassPropertyTypesVisitor.java#L58) was a caller of `getNode` and thus is broken. I don't think there's anything we can do in this case, since this is an essential/unavoidable change, but this breaks the compatibility policy for version 6.x, right?

Internal side: I tried to solve this issue within the constraint that `ApexNode` expects to wrap just one internal node. If we do decide to ignore this constraint, I definitely see an advantage to your solution over the fake nodes solution I proposed, as it would allow for a more natural interface to the list of nodes. My current idea is to create a sibling of `AbstractApexNode` (like you mentioned) that holds a list of Summit nodes rather than a single one, allowing a class like `ASTModifierNode` to have a signature like `ASTModifierNode extends AbstractApexListNode<Modifier>` with the `List<Modifier>` accessed through `this.nodes`. Please let me know if I'm understanding your idea correctly. Could you elaborate more on the exception & dummy type you mentioned for `getNode`?

Reply:
Thanks for testing `pmd-visualforce`. You're right: it does use Jorje. I think that's a separable enough issue that we can move to the next steps without it. I'll take a look in the next week and try to scope it out.

Yup, I don't believe there is anything (reasonable) that could be done to maintain the strict compatibility, thus I have/had no expectation that `experimental-apex-parser` will ever be merged into `master` except though 7.x.

After fielding some general questions from Bogdan, I'm feeling more confident that no class or interface that is referenced outside of the `ast/` package should have a generic type parameter that exposes the wrapped Nodes (and furthermore assumes a 1:1 mapping). Specifically, `ApexNode<T extends wrapped.Node>`. In practice, most of the external references are `ApexNode<?>`... but why even have it?

I would suggest migrating `ApexNode<T extends wrapped.Node>` to `ApexNode` in two phases:
* Instantiate all of its uses with a dummy type, like `Void`.  For example, `class ASTUserClass implements ApexNode<Void>`. All of the uses of `ApexNode<?>` stay ok... and more specific ones were already problem.
* Optionally, have an internal base class for the subset of nodes with a 1:1 mapping, if this makes `this.node` convenient. Or don't have it, and each class stores what it wraps. Whatever you think works best.
* Sometime in the future (... maybe PMD 7.x), remove the generic type parameter.

Reply:
Resolved: will be addressed in separate PR